File: not-proven/proven-syslog/ffi/zig/src/syslog.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// syslog.zig — Zig FFI implementation of proven-syslog.
//
// Implements the Syslog collector primitive (RFC 5424) with:
//   - Slot-based collector management (up to 64 concurrent collectors)
//   - Message ingestion with facility (0-23) and severity (0-7) tracking
//   - Priority computation (facility * 8 + severity)
//   - Transport protocol management (UDP/514, TCP/514, TLS/6514)
//   - Severity-based message filtering
//   - Message and drop counters
//   - Thread-safe via mutex on global state
//
// Tag values MUST match:
//   - Idris2 ABI (src/abi/Types.idr)
//   - C header   (generated/abi/syslog.h)

const std = @import("std");

// ── Enums (matching Idris2 SyslogABI.Types tag assignments exactly) ─────

/// Severity — matches severityToTag (RFC 5424 Section 6.2.1)
pub const Severity = enum(u8) {
    emergency = 0,
    alert = 1,
    critical = 2,
    err = 3,
    warning = 4,
    notice = 5,
    informational = 6,
    debug = 7,
};

/// Facility — matches facilityToTag (RFC 5424 Section 6.2.1)
pub const Facility = enum(u8) {
    kern = 0,
    user = 1,
    mail = 2,
    daemon = 3,
    auth = 4,
    syslog_f = 5,
    lpr = 6,
    news = 7,
    uucp = 8,
    cron = 9,
    auth_priv = 10,
    ftp = 11,
    ntp = 12,
    audit = 13,
    alert_f = 14,
    clock = 15,
    local0 = 16,
    local1 = 17,
    local2 = 18,
    local3 = 19,
    local4 = 20,
    local5 = 21,
    local6 = 22,
    local7 = 23,
};

/// Transport — matches transportToTag
pub const Transport = enum(u8) {
    udp_514 = 0,
    tcp_514 = 1,
    tls_6514 = 2,
};

/// SyslogError — error codes for FFI operations
pub const SyslogError = enum(u8) {
    ok = 0,
    invalid_slot = 1,
    not_active = 2,
    invalid_facility = 3,
    invalid_severity = 4,
    invalid_transport = 5,
    filtered = 6,
};

// ── Collector Context ───────────────────────────────────────────────────

const CollectorCtx = struct {
    /// Whether this slot is in use.
    active: bool,
    /// Transport protocol.
    transport: Transport,
    /// Total messages ingested (including filtered).
    message_count: u32,
    /// Messages dropped by severity filter.
    dropped_count: u32,
    /// Last facility tag (255 = none).
    last_facility: u8,
    /// Last severity tag (255 = none).
    last_severity: u8,
    /// Minimum severity filter (7 = Debug, accept all by default).
    min_severity: u8,
};

// ── Global state (slot-based, mutex-protected) ──────────────────────────

const MAX_CONTEXTS: usize = 64;

const empty_ctx: CollectorCtx = .{
    .active = false,
    .transport = .udp_514,
    .message_count = 0,
    .dropped_count = 0,
    .last_facility = 255,
    .last_severity = 255,
    .min_severity = 7, // Accept all (Debug is least severe)
};

var contexts: [MAX_CONTEXTS]CollectorCtx = [_]CollectorCtx{empty_ctx} ** MAX_CONTEXTS;
var mutex: std.Thread.Mutex = .{};

// ── Helpers ─────────────────────────────────────────────────────────────

/// Validate and return a pointer to an active context, or null.
fn getActive(slot: c_int) ?*CollectorCtx {
    if (slot < 0 or slot >= MAX_CONTEXTS) return null;
    const idx: usize = @intCast(slot);
    if (!contexts[idx].active) return null;
    return &contexts[idx];
}

// ── ABI version ─────────────────────────────────────────────────────────

/// ABI version — must match SyslogABI.Foreign.abiVersion (currently 1).
pub export fn syslog_abi_version() callconv(.c) u32 {
    return 1;
}

// ── Lifecycle ───────────────────────────────────────────────────────────

/// Create a new syslog collector context.
/// Returns slot index (0-63) or -1 if no slots available or invalid params.
pub export fn syslog_create(transport: u8) callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    // Validate transport (0-2)
    if (transport > 2) return -1;

    for (&contexts, 0..) |*ctx, i| {
        if (!ctx.active) {
            ctx.* = empty_ctx;
            ctx.active = true;
            ctx.transport = @enumFromInt(transport);
            return @intCast(i);
        }
    }
    return -1;
}

/// Destroy a collector context, freeing its slot.
pub export fn syslog_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();

    if (slot < 0 or slot >= MAX_CONTEXTS) return;
    const idx: usize = @intCast(slot);
    contexts[idx].active = false;
}

// ── State queries ───────────────────────────────────────────────────────

/// Get the Transport tag for a slot.
/// Returns UDP514 (0) for invalid/inactive slots.
pub export fn syslog_get_transport(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 0;
    return @intFromEnum(ctx.transport);
}

/// Get the total number of messages ingested.
pub export fn syslog_get_message_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 0;
    return ctx.message_count;
}

/// Get the last facility tag (255 = none).
pub export fn syslog_get_last_facility(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 255;
    return ctx.last_facility;
}

/// Get the last severity tag (255 = none).
pub export fn syslog_get_last_severity(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 255;
    return ctx.last_severity;
}

/// Get the last computed priority value (facility * 8 + severity).
/// Returns 0xFFFFFFFF for invalid/inactive slots or if no message ingested.
pub export fn syslog_get_last_priority(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 0xFFFFFFFF;
    if (ctx.last_facility == 255 or ctx.last_severity == 255) return 0xFFFFFFFF;
    return @as(u32, ctx.last_facility) * 8 + @as(u32, ctx.last_severity);
}

/// Get the minimum severity filter tag.
pub export fn syslog_get_min_severity(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 7;
    return ctx.min_severity;
}

/// Get the count of messages dropped by severity filter.
pub export fn syslog_get_dropped_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 0;
    return ctx.dropped_count;
}

// ── Operations ──────────────────────────────────────────────────────────

/// Set the transport protocol.
/// Returns SyslogError tag.
pub export fn syslog_set_transport(slot: c_int, t: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return @intFromEnum(SyslogError.invalid_slot);

    if (t > 2) {
        return @intFromEnum(SyslogError.invalid_transport);
    }

    ctx.transport = @enumFromInt(t);
    return @intFromEnum(SyslogError.ok);
}

/// Set the minimum severity filter.
/// Messages with severity > min_severity (less severe) will be dropped.
/// Returns SyslogError tag.
pub export fn syslog_set_min_severity(slot: c_int, sev: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return @intFromEnum(SyslogError.invalid_slot);

    if (sev > 7) {
        return @intFromEnum(SyslogError.invalid_severity);
    }

    ctx.min_severity = sev;
    return @intFromEnum(SyslogError.ok);
}

/// Ingest a syslog message with given facility and severity.
/// Messages less severe than min_severity are counted but marked as dropped.
/// Returns SyslogError tag (Filtered if dropped).
pub export fn syslog_ingest(slot: c_int, fac: u8, sev: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return @intFromEnum(SyslogError.invalid_slot);

    if (fac > 23) {
        return @intFromEnum(SyslogError.invalid_facility);
    }

    if (sev > 7) {
        return @intFromEnum(SyslogError.invalid_severity);
    }

    ctx.last_facility = fac;
    ctx.last_severity = sev;
    ctx.message_count += 1;

    // Filter: lower numeric severity = more severe.
    // Drop if message severity > min_severity (less severe than threshold).
    if (sev > ctx.min_severity) {
        ctx.dropped_count += 1;
        return @intFromEnum(SyslogError.filtered);
    }

    return @intFromEnum(SyslogError.ok);
}

// ── Stateless priority computation ──────────────────────────────────────

/// Compute the syslog priority value: facility * 8 + severity.
/// Returns 0xFFFFFFFF for invalid facility or severity.
pub export fn syslog_compute_priority(fac: u8, sev: u8) callconv(.c) u32 {
    if (fac > 23 or sev > 7) return 0xFFFFFFFF;
    return @as(u32, fac) * 8 + @as(u32, sev);
}

File: bindings/elixir/lib/proven_servers/syslog.ex
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
defmodule ProvenServers.Syslog do
  @moduledoc """
  Syslog protocol types for the proven-servers ABI.
  
  Mirrors the Idris2 module `SyslogABI.Types` and its type definitions:
  - `Severity`   — syslog severity levels (8 constructors, tags 0-7)
  - `Facility`   — syslog facility codes (24 constructors, tags 0-23)
  - `Transport`  — syslog transport mechanisms (3 constructors, tags 0-2)
  
  Severity and facility values match RFC 5424 numeric codes.
  All discriminant values match the Idris2 ABI tag definitions exactly.

  All tag values match the Idris2 ABI definitions exactly.
  """

  @doc "Standard syslog UDP port (RFC 5426)."
  @spec syslog_udp_port() :: non_neg_integer()
  def syslog_udp_port, do: 514

  @doc "Standard syslog TCP port (RFC 6587)."
  @spec syslog_tcp_port() :: non_neg_integer()
  def syslog_tcp_port, do: 514

  @doc "Syslog over TLS port (RFC 5425)."
  @spec syslog_tls_port() :: non_neg_integer()
  def syslog_tls_port, do: 6514

  # ===========================================================================
  # Severity (tags 0-7)
  # ===========================================================================

  @typedoc """
  Severity types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type severity ::
          :emergency
          | :alert
          | :critical
          | :error
          | :warning
          | :notice
          | :informational
          | :debug

  @severity_tags %{
    emergency: 0,
    alert: 1,
    critical: 2,
    error: 3,
    warning: 4,
    notice: 5,
    informational: 6,
    debug: 7,
  }

  @tag_to_severity Map.new(@severity_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `Severity` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..7, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Syslog.severity_from_tag(0)
      {:ok, :emergency}
  """
  @spec severity_from_tag(non_neg_integer()) :: {:ok, severity()} | :error
  def severity_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 7 do
    {:ok, Map.fetch!(@tag_to_severity, tag)}
  end

  def severity_from_tag(_tag), do: :error

  @doc """
  Encode a `Severity` to the C-ABI tag value.
  """
  @spec severity_to_tag(severity()) :: non_neg_integer()
  def severity_to_tag(val) when is_map_key(@severity_tags, val) do
    Map.fetch!(@severity_tags, val)
  end

  @doc """
  All `Severity` variants in tag order.
  """
  @spec all_severitys() :: [severity()]
  def all_severitys do
    [
      :emergency, :alert, :critical, :error, :warning, :notice, :informational,
      :debug
    ]
  end

  # ===========================================================================
  # Facility (tags 0-23)
  # ===========================================================================

  @typedoc """
  Facility types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type facility ::
          :kern
          | :user
          | :mail
          | :daemon
          | :auth
          | :syslog
          | :lpr
          | :news
          | :uucp
          | :cron
          | :auth_priv
          | :ftp
          | :ntp
          | :audit
          | :alert
          | :clock
          | :local0
          | :local1
          | :local2
          | :local3
          | :local4
          | :local5
          | :local6
          | :local7

  @facility_tags %{
    kern: 0,
    user: 1,
    mail: 2,
    daemon: 3,
    auth: 4,
    syslog: 5,
    lpr: 6,
    news: 7,
    uucp: 8,
    cron: 9,
    auth_priv: 10,
    ftp: 11,
    ntp: 12,
    audit: 13,
    alert: 14,
    clock: 15,
    local0: 16,
    local1: 17,
    local2: 18,
    local3: 19,
    local4: 20,
    local5: 21,
    local6: 22,
    local7: 23,
  }

  @tag_to_facility Map.new(@facility_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `Facility` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..23, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Syslog.facility_from_tag(0)
      {:ok, :kern}
  """
  @spec facility_from_tag(non_neg_integer()) :: {:ok, facility()} | :error
  def facility_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 23 do
    {:ok, Map.fetch!(@tag_to_facility, tag)}
  end

  def facility_from_tag(_tag), do: :error

  @doc """
  Encode a `Facility` to the C-ABI tag value.
  """
  @spec facility_to_tag(facility()) :: non_neg_integer()
  def facility_to_tag(val) when is_map_key(@facility_tags, val) do
    Map.fetch!(@facility_tags, val)
  end

  @doc """
  All `Facility` variants in tag order.
  """
  @spec all_facilitys() :: [facility()]
  def all_facilitys do
    [
      :kern, :user, :mail, :daemon, :auth, :syslog, :lpr, :news, :uucp,
      :cron, :auth_priv, :ftp, :ntp, :audit, :alert, :clock, :local0,
      :local1, :local2, :local3, :local4, :local5, :local6, :local7,
    ]
  end

  @doc """
  Decode from an ABI tag value (RFC 5424 facility code).

  Encode to the ABI tag value (RFC 5424 facility code).

  Whether this is a local-use facility (Local0-Local7).
        (self as u8) >= 16

  Whether this is a security-related facility.
  """
  @spec is_security?(facility()) :: boolean()
  def is_security?(val) when val in [:auth, :auth_priv, :audit], do: true
  def is_security?(_val), do: false

  # ===========================================================================
  # Transport (tags 0-2)
  # ===========================================================================

  @typedoc """
  Transport types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type transport :: :udp514 | :tcp514 | :tls6514

  @transport_tags %{
    udp514: 0,
    tcp514: 1,
    tls6514: 2,
  }

  @tag_to_transport Map.new(@transport_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `Transport` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..2, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Syslog.transport_from_tag(0)
      {:ok, :udp514}
  """
  @spec transport_from_tag(non_neg_integer()) :: {:ok, transport()} | :error
  def transport_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 2 do
    {:ok, Map.fetch!(@tag_to_transport, tag)}
  end

  def transport_from_tag(_tag), do: :error

  @doc """
  Encode a `Transport` to the C-ABI tag value.
  """
  @spec transport_to_tag(transport()) :: non_neg_integer()
  def transport_to_tag(val) when is_map_key(@transport_tags, val) do
    Map.fetch!(@transport_tags, val)
  end

  @doc """
  All `Transport` variants in tag order.
  """
  @spec all_transports() :: [transport()]
  def all_transports, do: [:udp514, :tcp514, :tls6514]

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  The port number used by this transport.
        match self {

  Whether this transport provides encryption.
  """
  @spec is_encrypted?(transport()) :: boolean()
  def is_encrypted?(val) when val in [:tls6514], do: true
  def is_encrypted?(_val), do: false

  @doc """
  Whether this transport provides reliable delivery.
  """
  @spec is_reliable?(transport()) :: boolean()
  def is_reliable?(val) when val in [:tcp514, :tls6514], do: true
  def is_reliable?(_val), do: false

end
